src/org/etools/j1939_84/controllers/part12/Part12Step12Controller.java

```diff
@@ -0,0 +1,211 @@
+
+                                                                 DM58RationalityFaultSpData packet = dm58.get();
+                                                                 // c. Fail, if expected unused bytes in DM58 are
+                                                                 // not padded with FFh
+                                                                 if (!areUnusedBytesPaddedWithFFh(packet)) {
+                                                                     addFailure(
+                                                                                "6.12.12.3.c - Unused bytes in DM58 are not padded with FFh in the response from "
+                                                                                        + module.getModuleName()
+                                                                                        + " for SP " + spn);
+                                                                 }
+                                                                 // d. Fail, if data returned is greater than FBh
+                                                                 // (for 1 byte SP), FBFFh (for 2 byte SP), or
+                                                                 // FBFFFFFFh (for 4 byte SP)
+                                                                 if (isGreaterThanFb(packet)) {
+                                                                     addFailure(
+                                                                                "6.12.12.d - Data returned is greater than 0xFB... threshold from "
+                                                                                        + module.getModuleName()
+                                                                                        + " for " + spn);
+
+                                                                 }
+                                                             })
+                                                             .collect(Collectors.toList());
+                      if (packets.isEmpty()) {
+                          addFailure("6.12.12.3.a - NACK received for DM7 PG from OBD ECU "
+                                  + module.getModuleName() + " for SP " + spn);
+                      }
+
+                  });
+
+        });
+
+        getDataRepository().getObdModuleAddresses().forEach(moduleAddress -> {
+            getDm58AndVerifyData(moduleAddress);
+        });
+
+    }
+
+    private void getDm58AndVerifyData(int moduleAddress) {
+
+        // 6.12.11.4 Actions3:
+        // a. DS DM7 with TID 245 (for DM58) using FMI 31 for first SP identified as not supporting DM58 in a DM24
+        // response In step 6.1.4.1 to the SP’s respective OBD ECU. (Use of an SP that supports test results is
+        // preferred when available).
+        var nonRatFaultSps = getDataRepository().getObdModules()
```

Comment:
This should be using only the module specified by the `moduleAddress` that’s passed in.